packages/core/mesh/network-manager/src/swarm/connection.ts

```diff
@@ -348,6 +352,18 @@ export class Connection {
     this._callbacks?.onClosed?.(err);
   }
 
+  private async _closeProtocol() {
```

Comment:
this addresses the non-reentrancy of synchronized destroy. 
if `this._protocol.stream.on('closed'` is triggered when `protocol.close()` is called we'll start waiting for the lock for `destroy()` which we actually hold.
use a trigger to signal the code in `destroy()` to continue

Reply:
`await protocol.close() -> teleport.close() -> extension.close() -> context.dispose() -> leaveSwarm() -> peer.destroy() -> connection.close()` 
I could fix it for this particular case, but it's very error-prone 🤷‍♂️